packages/flutter/lib/src/cupertino/dialog.dart

```diff
@@ -196,29 +199,153 @@ class CupertinoAlertDialog extends StatelessWidget {
+      ),
+      child: actionSection,
+    );
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return new Center(
+      child: new Container(
+        width: _kCupertinoDialogWidth,
+        padding: const EdgeInsets.symmetric(vertical: _kEdgePadding),
+        child: new CustomMultiChildLayout(
+          delegate: new _CupertinoAlertDialogLayoutDelegate(
+            actionsCount: actions.length,
+          ),
+          children: <Widget>[
+            new LayoutId(
+              id: _AlertDialogSection.blur,
+              child: _buildBlurBackground(),
+            ),
+            new LayoutId(
+              id: _AlertDialogSection.content,
+              child: _buildContent(),
+            ),
+            new LayoutId(
+              id: _AlertDialogSection.actions,
+              child: _buildActions(),
+            ),
+          ],
         ),
       ),
     );
   }
 }
 
+enum _AlertDialogSection {
+  blur,
+  content,
+  actions,
+}
+
+class _CupertinoAlertDialogLayoutDelegate extends MultiChildLayoutDelegate {
+
+  final int actionsCount;
+
+  _CupertinoAlertDialogLayoutDelegate({
+    this.actionsCount,
+  });
+
+  @override
+  void performLayout(Size size) {
+    // TODO: @mattcarroll, handle 2 buttons that need to stack due to long text
+    double minActionSpace = 0.0;
+    if (actionsCount > 0 && actionsCount <= 2) {
+      minActionSpace = _kButtonHeight;
+    } else {
+      minActionSpace = 1.5 * _kButtonHeight;
+    }
+
+    // Size alert dialog content.
+    final Size maxContentSize = new Size(size.width, size.height - minActionSpace);
+    final Size contentSize = layoutChild(
+        _AlertDialogSection.content,
+        new BoxConstraints.loose(maxContentSize),
+    );
+
+    // Size alert dialog actions.
+    final Size maxActionSize = new Size(
+        size.width,
+        size.height - contentSize.height,
+    );
+    final Size actionsSize = layoutChild(
+        _AlertDialogSection.actions,
+        new BoxConstraints.loose(maxActionSize),
+    );
+
+    // Calculate overall dialog height.
+    final double dialogHeight = contentSize.height + actionsSize.height;
+
+    // Size blur background.
+    final Size backgroundSize = new Size(size.width, dialogHeight);
+    layoutChild(
+      _AlertDialogSection.blur,
+      new BoxConstraints.loose(backgroundSize),
+    );
+
+    // Layout the blur, content, and the actions.
+    final double dialogTop = (size.height - dialogHeight) / 2;
+    positionChild(
+      _AlertDialogSection.blur,
+      new Offset(
+        0.0,
+        dialogTop,
+      ),
+    );
+    positionChild(
+        _AlertDialogSection.content,
+        new Offset(
+          0.0,
+          dialogTop,
+        ),
+    );
+    positionChild(
+        _AlertDialogSection.actions,
+        new Offset(
+          0.0,
+          dialogTop + contentSize.height,
+        ),
+    );
+  }
+
+  @override
+  bool shouldRelayout(MultiChildLayoutDelegate oldDelegate) {
+    return false;
```

Comment:
why false? what if the actionsCount changes?

Reply:
I'll change it to check action count.  This invalidation will need some thought to get it right.  Because it's not just action count, it's text length and stacking.  I'll add a todo for that.

Reply:
we should make sure to get this right before landing.

i don't see the two things you mentioned in the list of constructor arguments, though, so how can they affect the algorithm?

Reply:
Yeah, this was a rough approach for the common case because there are a few issues that come together here.  First, I'm not sure that the CupertinoDialogAction handles text wrapping correctly.  Maybe it does, but I need to check.  Second, we're not stacking buttons correctly.  Currently we're only stacking when 3+ buttons are used.  But, actually, based on the desired width of the buttons, there are times when we should stack 2.  That as-of-yet-unimplemented stacking behavior impacts this invalidation.  There is also an open ended question as to what the button layout logic means with regard to arbitrary Widgets that a developer chooses to pass instead of CupertinoDialogActions.

Due to those TODOs and unknowns I figured we should take the implest approach possible here that handles the typical cases.  Which I think this does.